### ISSUE TITLE
StartSpanConfig.Parent will not work with anything that satisfies SpanContext

### ISSUE THREAD

**Issue by drewlesueur** [DataDog/dd-trace-go#733]
I am wondering if I am reading the documentation wrong, but it seems that when using a [ddtrace.StartSpanConfig](https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/ddtrace#StartSpanConfig), I should be able to use anything that satisfies [ddtrace.SpanContext](https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/ddtrace#SpanContext) as `Parent`.
So when using [tracer.ChildOf](https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer#ChildOf) I should be able to also pass in anything that satisfies ddtrace.SpanContext.

However the code for StartSpan seems to NOT care about the ddtrace.SpanContext interface but only the internal *spanContext concrete type (https://github.com/DataDog/dd-trace-go/blob/v1.26.0/ddtrace/tracer/tracer.go#L249)
`if ctx, ok := opts.Parent.(*spanContext); ok {`

This is not allowing me to manually do distributed tracing by trying to use a custom type that satisfies SpanContext.
Am I understanding wrong how it's supposed to work?

Summary: It seems documentation shows that I should be able to use anything that implements ddtrace.SpanContext as an argument to tracer.ChildOf, but in practice it looks like I can't.

Example program:
```
package main

import (
    "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
    "log"
)

// Context is a concrete struct that satisfies ddtrace.SpanContext
type Context struct {
    InternalSpanID  uint64
    InternalTraceID uint64
}

// SpanID returns the SpanID for this context
// or 0 if nil. This function helps satisfy the ddtrace.SpanContext interface.
func (t *Context) SpanID() uint64 {
    if t == nil {
        return 0
    }
    return t.InternalSpanID
}

// TraceID returns the TraceID for this context 
// or 0 if nil. This function helps satisfy the ddtrace.SpanContext interface.
func (t *Context) TraceID() uint64 {
    if t == nil {
        return 0
    }
    return t.InternalTraceID
}

func (t *Context) ForeachBaggageItem(handler func(k, v string) bool) {
    // Unimplemented because it's not integral to the example.
}

func main() {
    tracer.Start(
        tracer.WithAgentAddr(""),
        tracer.WithService("test"),
    )   
    defer tracer.Stop()
    
    tracingContext := &Context{
        InternalTraceID: 1, 
        InternalSpanID:  2, 
    }   
    span := tracer.StartSpan("test_span", tracer.ChildOf(tracingContext))
    log.Printf("span: %v", span)
    // The span's TraceID is not 1, but based on the documentation, I think it should be.
    span.Finish() 
}   
```

**Comment by knusbaum** [comment]
Hey @drewlesueur. Thanks for the question.
### Why it doesn't work
The interfaces in `"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"` define how a tracer should behave.

We have 3 implementations of the tracer described by `"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"`:
- `"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"` - The standard datadog tracer
- `"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/opentracer"` - A tracer for use with opentracing
- `"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"` - A tracer for testing.

Even though the tracers conform to the interfaces in `"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"`, they *may not* accept every implementation of those interfaces. The standard tracer, as you discovered, only understands its own internal `SpanContext`. 

### How to implement custom distributed tracing
For a solution to your problem, have a look at the [standard tracer documentation](https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer)
>To make use of distributed tracing, a span's context may be injected via a carrier into a transport (HTTP, RPC, etc.) to be extracted on the other end and used to create spans that are direct descendants of it. A couple of carrier interfaces which should cover most of the use-case scenarios are readily provided, such as HTTPCarrier and TextMapCarrier. Users are free to create their own, which will work with our propagation algorithm as long as they implement the TextMapReader and TextMapWriter interfaces. An example alternate implementation is the MDCarrier in our gRPC integration.

tl;dr: 
Use one of the existing carriers, such as [`TextMapCarrier`](https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer#TextMapCarrier):
```
	s := tracer.StartSpan("parent.span")
	carrier := make(map[string]string)
	// Inject the span's context into carrier
	err := tracer.Inject(s.Context(), tracer.TextMapCarrier(carrier))
	//...
	// Send the carrier where you want. Across the network, wherever.
	// ...
	// On the other end:
	sctx, err := tracer.Extract(tracer.TextMapCarrier(carrier))
	span := tracer.StartSpan("child.span", tracer.ChildOf(sctx))
```

You may also implement a [TextMapReader](https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer#TextMapReader) and [TextMapWriter](https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer#TextMapWriter) if none of the existing carriers work for you.

**Comment by drewlesueur** [comment]
Thank you very much for the response.
Using Extract with a struct that implements TestMapReader worked great.
